File: forms/integumentary-assessment/full-stack-with-loco-tera-htmx-alpine/src/engine/integumentary_grader.rs
use super::flagged_issues::detect_additional_flags;
use super::rules::all_rules;
use super::types::{AssessmentData, FiredRule, GradingResult};
use super::utils::classify_braden_score;

/// Pure function: grades an integumentary assessment using the Braden Scale.
///
/// Returns a [`GradingResult`] containing:
/// - the Braden total score (6-23; defaults to 23 / "no-risk" if no
///   subscales have been answered, mirroring the reference engine)
/// - the derived risk level slug
/// - per-subscale audit trail (`fired_rules`)
/// - clinician-facing additional flags (urgent > high > medium > low)
pub fn grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();

    let mut fired_rules: Vec<FiredRule> = Vec::new();
    let mut total_score: i32 = 0;
    let mut answered_count: i32 = 0;

    for rule in all_rules() {
        let score = (rule.evaluate)(data);
        if score > 0 {
            answered_count += 1;
            total_score += score;
            fired_rules.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                description: rule.description.to_string(),
                score,
                max_score: rule.max_score,
            });
        }
    }

    let braden_score = if answered_count == 0 { 23 } else { total_score };
    let risk_level = classify_braden_score(braden_score);
    let additional_flags = detect_additional_flags(data);

    GradingResult {
        braden_score,
        risk_level,
        answered_count,
        fired_rules,
        additional_flags,
        timestamp,
    }
}

File: forms/integumentary-assessment/front-end-with-html/js/integumentary-grader.js
import { bradenRules } from './rules.js';

// Braden Scale grader. Pure functions: take an `AssessmentData` object,
// return the total Braden score (6-23), the `RiskLevel`, and the list of
// fired subscales. Unanswered subscales are excluded from the total but
// tracked separately.
//
// Risk-level cutoffs (Braden total score, lower = higher risk):
//   <=  9  -> Very High Risk
//   10-12  -> High Risk
//   13-14  -> Moderate Risk
//   15-18  -> Mild Risk
//   19-23  -> No Risk

/**
 * @typedef {import('./types.js').AssessmentData} AssessmentData
 * @typedef {import('./types.js').RiskLevel} RiskLevel
 * @typedef {import('./types.js').FiredRule} FiredRule
 */

// Wrapped in an IIFE; published via window.IntegumentaryAssessment.

/**
 * Classify a numeric Braden score (6-23) into a risk level.
 * @param {number} score
 * @returns {RiskLevel}
 */
function classifyBradenScore(score) {
  if (score <= 9) return 'very-high-risk';
  if (score <= 12) return 'high-risk';
  if (score <= 14) return 'moderate-risk';
  if (score <= 18) return 'mild-risk';
  return 'no-risk';
}

/**
 * Friendly label for a RiskLevel.
 * @param {RiskLevel} level
 */
function riskLevelLabel(level) {
  switch (level) {
    case 'very-high-risk': return 'Very High Risk';
    case 'high-risk': return 'High Risk';
    case 'moderate-risk': return 'Moderate Risk';
    case 'mild-risk': return 'Mild Risk';
    case 'no-risk': return 'No Risk';
    default: return '';
  }
}

/**
 * CSS class hint for the risk-level badge.
 * @param {RiskLevel} level
 */
function riskLevelClass(level) {
  switch (level) {
    case 'very-high-risk': return 'risk-very-high';
    case 'high-risk': return 'risk-high';
    case 'moderate-risk': return 'risk-moderate';
    case 'mild-risk': return 'risk-mild';
    case 'no-risk': return 'risk-no';
    default: return '';
  }
}

/**
 * Evaluate the 6-subscale Braden Scale against the supplied assessment data
 * and produce the total score plus per-subscale audit trail.
 *
 * If no subscales are answered, the total is reported as 23 (no risk) so
 * an empty submission classifies as "No Risk" rather than the highest-risk
 * extreme of 6. Mirrors the SvelteKit reference engine.
 *
 * @param {AssessmentData} data
 * @returns {{ bradenScore: number, riskLevel: RiskLevel, answeredCount: number, firedRules: FiredRule[] }}
 */
function calculateBraden(data) {
  /** @type {FiredRule[]} */
  const firedRules = [];
  let totalScore = 0;
  let answeredCount = 0;

  for (const rule of bradenRules) {
    try {
      const score = rule.evaluate(data);
      if (score > 0) {
        answeredCount++;
        totalScore += score;
        firedRules.push({
          id: rule.id,
          category: rule.category,
          description: rule.description,
          score,
          maxScore: rule.maxScore
        });
      }
    } catch (e) {
      console.warn(`Braden rule ${rule.id} evaluation failed:`, e);
    }
  }

  const bradenScore = answeredCount === 0 ? 23 : totalScore;
  const riskLevel = classifyBradenScore(bradenScore);

  return { bradenScore, riskLevel, answeredCount, firedRules };
}

export { classifyBradenScore, riskLevelLabel, riskLevelClass, calculateBraden };
